xds/internal/server/rds_handler.go

```diff
@@ -131,3 +130,33 @@ func (rh *rdsHandler) close() {
 		cancel()
 	}
 }
+
+// rdsWatcher implements the xdsresource.RouteConfigWatcher interface and is
+// passed to the WatchRouteConfig API.
+type rdsWatcher struct {
+	parent *rdsHandler
+	logger *internalgrpclog.PrefixLogger
+	name   string
+}
+
+func (rw *rdsWatcher) OnUpdate(update *xdsresource.RouteConfigResourceData) {
+	if rw.logger.V(2) {
+		rw.logger.Infof("RDS watch for resource %q received update: %#v", rw.name, update.Resource)
+	}
+	rw.parent.handleRouteUpdate(rw.name, update.Resource)
+}
+
+func (rw *rdsWatcher) OnError(err error) {
+	if rw.logger.V(2) {
+		rw.logger.Infof("RDS watch for resource %q reported error: %v", rw.name, err)
+	}
+	drainAndPush(rw.parent.updateChannel, rdsHandlerUpdate{err: err})
+}
+
+func (rw *rdsWatcher) OnResourceDoesNotExist() {
+	if rw.logger.V(2) {
+		rw.logger.Infof("RDS watch for resource %q reported resource-does-not-exits error: %v", rw.name)
+	}
+	err := xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "resource name %q of type RouteConfiguration not found in received response", rw.name)
```

Comment:
Listener and RouteConfiguration are the proto types? (you log type listener in other handler)

Reply:
Listener and RouteConfiguration are the xDS resource types, and the proto messages are 1:1.

> you log type listener in other handler

As I mentioned in the other comment, the only place where I refer to it as `Listener` is when I'm talking about the Listener resource.

Reply:
Alright sounds good.